# Statistics

In [27]:
import numpy as np

# Probability

Q1) In a group of 50 marbles, 15 are clear and the remaining 35 has either red, blue or both. a) If 20 are red and 25 are blue, how many are red and blue?

What is the probability of drawing; b) a red marble, c) a blue marble and d) a red and blue marble, e) red OR blue or multi-coloured marble at random from a bag?

In [15]:
#Sampling probability

#red = (20-x)
#blue = (25-x)
#35 = x+red+blue
x = 10

t = 15.+10.+20.+25.
p_red = 20./t
print p_red
p_blue = 25./t
print p_blue
p_both = 10./t
print p_both
p_both = 55./t
print p_both

0.285714285714
0.357142857143
0.142857142857
0.785714285714


A1a) 10, b) 0.29, c) 0.36, d) 0.14, e) 0.79

Q2) The S2DS 2016 cohort contained 90 people. The ages are distributed normally and have a mean age of 30 with a standard devation of 5. 

A student is selected at random. What is the probability that 

a) their age is greater than 35.

b) their age is less than 22.

c) thier age is between 22 and 35. 

d) Approximately how many students are older than 35?

In [16]:
#Normal distributions

(35.-30.)/5.
z_1 = .8413
p_35 = 1-z_1
(22.-30.)/5.
z_1_6 = .0548
p_22 = z_1_6

p_range = z_1-z_1_6
n_35 = p_35*90.

print p_35,p_22,p_range,n_35

0.1587 0.0548 0.7865 14.283


A2) a) 0.1587, b) 0.0548, c) 0.7865, d) 14.

Q3) A media firm identify that 85% of users of their app are using it to read sports related news articles. Out of a random sample of 100 users, find the probability that a) exactly three quarters of the sample are reading sports articles?

In [17]:
#Combination probability

import math
def C_n_r(n,r):
    N = math.factorial(n)
    N_R = math.factorial(n-r)
    R = math.factorial(r)
    return N/(N_R*R)

print C_n_r(100,75)*((0.85)**75.)*((1-0.85)**(100.-75.))

0.00311492186159


A3) 0.00311

Q) What is the probability that the total of two dice will be greater than 7, given that the first die is a 4?

In [18]:
#Conditional prob.

#4,4 - 4,5 - 4,6

p1 = 1./6.
p1_2 = 4./36.

print (p1_2)/p1



0.666666666667


A) 2/3

Q4) Calculate the probability being dealt a straight flush (5 consecuative cards of the same suit) off of the top of a dec of cards?

In [19]:
#Combination probability

deck = C_n_r(52,5)
SF_combos = 40

print 40./deck

1.53907716933e-05


A) 1.53907716933e-05

Q4) Each Pivigo team is composed of 4 people drawn randomly from a pool of 30 women and 60 men. What is the probability that the teams are composed of ; a) two women and two men, b) four men, c) at least 1 women. 

In [20]:
#Combination probability

A1 = float(C_n_r(30.,2.))
A2 = float(C_n_r(60.,2.))
T = float(C_n_r(90.,4.))
B1 = float(C_n_r(60.,4.))
B2 = float(C_n_r(30.,0.))

W1 = float(C_n_r(60.,3.))
W2 = float(C_n_r(30.,1.))
W5 = float(C_n_r(60.,1.))
W6 = float(C_n_r(30.,3.))
W7 = float(C_n_r(60.,0.))
W8 = float(C_n_r(30.,4.))

print A1*A2/T
print B1*B2/T
print ((W1*W2/T)+(A1*A2/T)+(W5*W6/T)+(W7*W8/T))

0.301327885598
0.190840994212
0.809159005788


A4) a) 0.3010, b) 0.1908, c) 0.8092

Q5)  A pharmacutical lab is testing a new anti-viral detection drug. They find a false positive rate of 0.2% and a false negative rate of 0.9%. The virus is estimated to be present in 1 in 50,000  people in the population. Calculate a) the instances of the drug failing to detect the virus and b) the instances of the drug incorrectly identifying the virus (hint: in format of 1 per X number of people).

In [21]:
#Classification

#Classification rates
FPR = 0.002
FNR = 0.009
#Infection rates
Pd = 1./50000.
Pn = 1.-Pd

#Instances of people with disease AND false negatives
FN = Pd*FNR
print 'Instances of virus not detected by the drug. 1:',(int(round(1/FN,-5)))
#Instances of people without the disease AND false positives
FP = Pn*FPR
print 'Instances of drug misclassifying people. 1:',(int(round(1/FP,0)))

Instances of virus not detected by the drug. 1: 5600000
Instances of drug misclassifying people. 1: 500


A5 a) 1:5,600,000, b) 1:500

Q6) What is the probability of rolling at least one 6 when rolling three dice?

In [22]:
#Probability

p = 1-((5./6.)**3.)
p

0.42129629629629617

A6) 0.421

Q)  0.04% of the population was inflected with swine flu. An experimental test has a 99.75% chance of returning a positive result if a subject has the virus. a) Calculate the probability that a randomly selected person tests positive. b) What is the probability that a person is infected, given that they tested positive?

In [36]:
def bayes(P_BA,P_A,P_B):
    return (P_BA*P_A)/P_B

def P_B(P_A,P_nA,P_B_A,P_B_nA):
    return (P_B_A*P_A)+(P_B_nA*P_nA)

P_I = 0.0004 #P(A)
P_P = 0.9975 #P(B|A)
P_nI = 1-P_I #P(-A)
P_N = 1-P_P  #P(B|-A)

print P_B(P_I,P_nI,P_P,P_N) #P(B)
print bayes(P_P,P_I,P_B(P_I,P_nI,P_P,P_N))

0.002898
0.13768115942


A) a) 0.0030, b) 0.138

Q) You are playing a game whereby you draw two cards at random from a deck of 52 playing cards with the aim of drawing matching suits. a) If you draw a spade, what is the probability that you will draw a second spade? b) What is the probability drawing two cards of the same suit?

A) a) 0.235, b) 0.018

The fraction of S2DS students born outside of the UK and currently working in the UK is 7/20. The total fraction born outside of the UK is 2/3. Find the probability that a student is currently working in the UK, given that they were born outside of the UK? 

In [33]:
0.35/0.666

0.5255255255255254

0.53

Q8) A fracking company is conducting exploratory drilling in three sites. Each site has 20% probability of proving profitable and each site it independant of the others. If sucessful, the pay off from a single site is $4 million, from two sites it is $3 million (each) and from all three sites it is $2.5 million each (due to excess supply lowering prices). Find the expected payoff.

In [24]:
#Probability

p_1 = 0.2*0.8*0.8
p_2 = 0.2*0.2*0.8
p_3 = 0.2*0.2*0.2

E = (3.*p_1*4.*1.)+(3.*p_2*3.*2)+(1.*p_3*2.5*3.)
E

2.1720000000000006

A8) $2.17 million

Q9) The leader of the 'British Democratic Party' has a mean public statisfaction rating of -8% with a standard deviation of 12 (assume a normal distribution). Find a) the probability that a person chosen at random will have ranked them greater than +15 and b) ,if the survey size is 1053, calculate the number of individuals whose rankings are considered 'outliers' with +/-3sigma. c) What were the upper and lower bounds of thier scores? Note:statisfaction rankings run from -100 to +100.



In [28]:
#Normal distribution/outliers

mean = -8
std = 12
upp = 15

def test(x,u,sigma,n):
    Z = (x-u)/(sigma/np.sqrt(n))
    return Z

t = test(mean,upp,std,1)
print t

sigma3 = 0.0013*2.
N = 1053*sigma3
print int(N)
print mean+(3*std),mean-(3*std)

-1.91666666667
2
28 -44


A9) a) 0.0276 b) 2, c) 28 and -44

Q10) A recruiter takes commission from on average 5 placements per month. Find the probability that a) they place at least one person per month,  and b) the probability that they we fail to place anyone in any given week. The manager want's to improve sales and sets a target of 6 or more placements per month. c) calculate the probability of meeting this. (Hint: use Possion's law and assume 4 weeks in a month).

In [29]:
#Poisson distributions

def poisson(mu,x):
    return ((np.exp(-mu)*(mu**x))/math.factorial(x))

mu = 5.
x = 0

print 1-poisson(mu,x)

mu_w = 5./4.
print poisson(mu_w,x)

print 1-(poisson(mu,0)+poisson(mu,1)+poisson(mu,2)+poisson(mu,3)+poisson(mu,4)+poisson(mu,5))


0.993262053001
0.28650479686
0.384039345167


A10) a) 0.993, b) 0.287, c) 0.384

Q11) The recruitment agency makes contact with 85 needy data scientists every month but only 7% are succesfully placed with a client. A) what is the probability of you sucessfully finding a job with this agency within a month? B) 3 other S2DS graduates join this agency. What is the probability that at least one of you find a job within a month?

In [30]:
#Binomial distribution

def binomial(n,r,p):
    q = 1-p
    return C_n_r(n,r)*(p**r)*(q**(n-r))

n = 85.+1.
x = 1.
p = 0.07

print binomial(n,x,p)

n = 85.+1.+3.

print binomial(n,1,p)+binomial(n,2,p)+binomial(n,3,p)+binomial(n,4,p)

0.0126080266562
0.243901253388


A11) a) 0.0126, b) 0.244 (???)

Q) You are playing the game perudo. They are 6 players and a total of 36 dice in play. The player to your left calls 13 sixes (recall that rolls of one are a wildcard that can equal any other chosen roll). What is the probability that there are exactly 13 sixes in play?

In [31]:
#Binomial

r = 13
n = 36
p = 1./3.

print binomial(n,r,p)

0.129147241582


A). 0.13

Q) You are given the following function:

y = log(x)+x^(1/2)-3

And an imput variable of x=5. a) calculate y. The standard deviation on x = 0.25. b) calculate the standard deivation on y numerically using a monte-carlo method. c) Compare this answer to that anayltical solution found using the propagation of error.




In [55]:
def func(x):
    return np.log(x)+(x**(0.5))-3

print func(5)

x = 5
std = 0.25
n = 100000

MCparts = np.random.normal(x,std,n)

MCstd = [func(i) for i in MCparts]

np.std(MCstd)

0.845505889934


0.1060205388094097

A] a) 0.846, b) 0.1060, c) 0.1059

In [11]:
def heating(in1,in2,in3):
    return 23+(0.056*in1)-(0.42*in2)+(1.5*in3)+(0.015*in2*in3)

heating(160,100,15)



34.96

# Sample statistics

In [32]:
import pandas as pd
import numpy as np
import sklearn.metrics as metrics
from scipy.stats import chisquare
from metrics import TS_metrics as metric_class
from scipy.stats import chisquare

df_predictions = pd.read_csv('Training_predictions.txt', sep=';')

Q1)Examine the income of users in the data set 'Training_predictions.txt'. Calculate a) the mean, b) median and c) modal income of a user.

In [71]:
#Counting statistics

mean = df_predictions['Income'].mean()
median = df_predictions['Income'].median()
mode = df_predictions['Income'].mode()
print 'Income: mean = {}, median = {}, mode = {}'.format(mean,median,mode)

Income: mean = 88210.0715, median = 85000.0, mode = 0    78000.0
dtype: float64


a) 88200 b) 85000 c) 78000

Q2) Find a) the standard deviation and b) variance of the age population age.

In [72]:
#Counting stats

std = df_predictions['Income'].std()
var = df_predictions['Income'].var()
print 'Income: std = {} Var = {}'.format(std,var)

Income: std = 37401.3903334 Var = 1398863998.87


A2) a) 37400 b) 1.4*10^9

Q3) We hypothesise that users older than 90 years are inhernetly wealthier. We propose a null hypothesis that states that "the 95% (or greater) confidence that the incomes of 90+ year old users IS consistent with being drawn randomly from the main sample".


Select a subsample of data for users over 90 years old. a) find its mean and b) calculate its 'one sample Z test statistic'. C) find the 95% condfidence interval of the 90 year old sample (hint: the Z score for 95% = 1.96sigma). D) Is the null hypothesis 'True' or 'False'

In [188]:
#Hypothesis testing

def test(x,u,sigma,n):
    Z = (x-u)/(sigma/np.sqrt(n))
    return Z

df_old = df_predictions.loc[df_predictions['Age'] > 90 ]
mean_old = df_old['Income'].mean()
n = len(df_old)
Z = test(mean_old,mean,std,n)

confidence_95 = 1.96*(std/np.sqrt(n))

print 'Old: Mean = {}+/- {}'.format(mean_old,confidence_95)
print '95% confidence = +/-1.96 Z-stat = {}'.format(Z)

Exception KeyboardInterrupt in 'zmq.backend.cython.message.Frame.__dealloc__' ignored


KeyboardInterrupt: 

A3] a) 88500
b) 1.32
c) 465
d) True

Q4] We hypothesise that customers of this company are more wealthy than the national average. We propose a null hypothesis that states that "there is a 95% or greater confidenece that customers could have been drawn randomly from the national population".

The mean national income is 26500. a) Calculate the "student's" t-stat value and b) find the 95% confidence interval of the customer database (hint: the t score for 95%=1.96sigma for inf. degrees of freedom). C) is the null hypothesis true or false?

In [106]:
#Hypothesis testing

mean_nat = 26500
n = len(df_predictions)
t = test(mean_nat,mean,std,n)
confidence_95 = 1.96*(std/np.sqrt(n))

print 'Customers: Mean = {}+/- {}'.format(mean_nat,confidence_95)
print '95% confidence = +/-1.96 t-stat = {}'.format(t)

Customers: Mean = 26500+/- 51.8356823919
95% confidence = +/-1.96 t-stat = -2333.36833931


A4] a) -16.5, b) 7331, c) False

Q5] Taking the standard deviation on the mean national income to be 6000 and assuming the distrbution is gaussian, use the two sample KS-test to compare the distributions of incomes national, and in the company. Find a) the KS-statistic and b) the probablity that each distribution is drawn from the same sample (p-value) - hint: as the population of the uk is 65 million but our data base is only 2 million you will need to normalise).

In [126]:
#Hypothesis testing

from scipy import stats
mu,sigma,n = 26500,6000,65.*(10.**6.)
UK_pop = np.random.normal(mu,sigma,n)

h,b = np.histogram(df_predictions['Income'],25,normed=True)
h_UK,b_UK = np.histogram(UK_pop,25,normed=True)

D,P = stats.ks_2samp(h,h_UK)
print 'KS-stats: D = ',D,' P = ',P

C:\Users\Damian\Anaconda2\lib\site-packages\ipykernel\__main__.py:3: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
  app.launch_new_instance()


KS-stats: D =  0.36  P =  0.0590270263056


A5] a) 0.36 b) 5.9%

# Metrics

Using the 'Modelling_predictions.txt' data set. Find the a) mean square error (MSE) b) mean absolute error (MAE) c) symmetric mean absolute percentage error (SMAPE) d) coefficent of determination (R^2) and e) Chi^2 value of the both the ORF and RRF Customer Value models, compared to the training Customer Values. Which is the best model?

In [1]:
#Validation

import pandas as pd
import numpy as np
import sklearn.metrics as metrics
from scipy.stats import chisquare
from metrics import TS_metrics as metric_class
from scipy.stats import chisquare

df_predictions = pd.read_csv('Training_predictions.txt', sep=';')
df_predictions.head()

#MSE
MSE_ORF = metrics.mean_squared_error(df_predictions.Customer_Value, 
                                  df_predictions.Predicted_Customer_Value_ORF)
MSE_RRF = metrics.mean_squared_error(df_predictions.Customer_Value, 
                                  df_predictions.Predicted_Customer_Value_RRF)
print 'a) MSE: ORF = {} RRF = {}'.format(MSE_ORF,MSE_RRF)
#MAE
MAE_ORF = metrics.mean_absolute_error(df_predictions.Customer_Value, 
                                  df_predictions.Predicted_Customer_Value_ORF)
MAE_RRF = metrics.mean_absolute_error(df_predictions.Customer_Value, 
                                  df_predictions.Predicted_Customer_Value_RRF)
print 'b) MAE: ORF = {} RRF = {}'.format(MAE_ORF,MAE_RRF)

M = metric_class(df_predictions.Customer_Value, 
            df_predictions.Customer_Value, 
            df_predictions.Predicted_Customer_Value_ORF)
SMAPE_ORF = M.SMAPE()
M = metric_class(df_predictions.Customer_Value, 
            df_predictions.Customer_Value, 
            df_predictions.Predicted_Customer_Value_RRF)
SMAPE_RRF = M.SMAPE()
print 'c)SMAPE: ORF = {} RRF = {}'.format(SMAPE_ORF,SMAPE_RRF)
R2_ORF = metrics.r2_score(df_predictions.Customer_Value, 
                                  df_predictions.Predicted_Customer_Value_ORF)       
R2_RRF = metrics.r2_score(df_predictions.Customer_Value, 
                                  df_predictions.Predicted_Customer_Value_RRF)   
print 'd) R2: ORF = {} RRF = {}'.format(R2_ORF,R2_RRF)
chi_score_ORF, pval_ORF = chisquare(df_predictions.Customer_Value, 
                                  df_predictions.Predicted_Customer_Value_ORF) 
chi_score_RRF, pval_RRF = chisquare(df_predictions.Customer_Value, 
                                  df_predictions.Predicted_Customer_Value_RRF) 
print 'e) Chi2: ORF = {} RRF = {}'.format(chi_score_ORF,chi_score_RRF)

a) MSE: ORF = 431.779410034 RRF = 6397.97143804
b) MAE: ORF = 9.23955673856 RRF = 39.8144096081
c)SMAPE: ORF = 10.5349253615 RRF = 43.9369328521
d) R2: ORF = 0.9437084819 RRF = 0.165889996982
e) Chi2: ORF = 4106432.51725 RRF = 157918924.366


A]

a) MSE: ORF = 432 RRF = 6400

b) MAE: ORF = 9.24 RRF = 39.8

c)SMAPE: ORF = 10.5 RRF = 43.9

d) R2: ORF = 0.944 RRF = 0.166

e) Chi2: ORF = 4110000 RRF = 158000000

ORF is the better model

In [42]:
res = df_predictions.Customer_Value - df_predictions.Predicted_Customer_Value_ORF

M = metric_class(res, 
            df_predictions.Customer_Value, 
            df_predictions.Predicted_Customer_Value_ORF)
M.forecast_accuracy_metric_std()

Mean =  -0.69263292248
Chi_square value:  183.297121963
P value:  1.57592553075e-40


(-0.69263292248000075, 183.2971219630123, 1.575925530747776e-40)

# Regression  modelling

Q1] Using the data set 'time_series_1.csv'. Fit a unweighted linear model to data and show that the coefficent of determination (R^2) is 97.4%.

In [29]:
#Validation

import pandas as pd
from sklearn import linear_model as lm

df = pd.read_csv("time_series_1.csv")

#Linear regression model

def linear_regression(df):
    '''
    Plot a weighted linear regression across a give data frame df
    '''
    x = df.TIME.values
    y = df.RATE.values

    length = len(x)
    x = x.reshape(length, 1)
    y = y.reshape(length, 1)

    regr = lm.LinearRegression()
    regr.fit(x, y)
    py = regr.predict(x)
    score_val = regr.score(x, y)

    return score_val

linear_regression(df)

0.97441739231778624

Q2] Using the data set 'time_series_1.csv'. Fit a weighted linear model to data and show that the coefficent of determination (R^2) is 97.3%.

In [30]:
#Validation

def Wlinear_regression(df):
    '''
    Plot a weighted linear regression across a give data frame df
    '''
    x = df.TIME.values
    y = df.RATE.values
    dy = df.ERROR.values

    length = len(x)
    x = x.reshape(length, 1)
    y = y.reshape(length, 1)
    Wdy = map(lambda x: x**(-2.),dy)# = dy.reshape(length, 1)

    regr = lm.LinearRegression()
    regr.fit(x, y,sample_weight=Wdy)
    py = regr.predict(x)
    score_val = regr.score(x, y,sample_weight=Wdy)
    chi_score, pval = chisquare(y, py)
    return score_val,chi_score,pval

Wlinear_regression(df)

(0.97285445949486071, array([ 90.94525624]), array([ 0.70574679]))

Q3] Using the data set 'time_series_2.csv'. Show that a wieghted polynomial model, of order 3, is a better model than a wieghted linear model, using Chi^2.

In [31]:
#Validation

df2 = pd.read_csv("time_series_2.csv")

def W_Polyfit(df,deg_val):
    '''
    Plot a weighted linear regression across a give data frame df
    '''
    x = df.TIME.values
    y = df.RATE.values
    dy = df.ERROR.values
    Wdy = map(lambda x: x**(-2.),dy)# = dy.reshape(length, 1)

    polyz = np.polyfit(x,y,deg=deg_val,w=Wdy)
    p = np.poly1d(polyz)
    xp = np.linspace(min(x), max(x), len(x))
    py = p(xp)
    
    chi_score, pval = chisquare(y, py)
    return chi_score,pval

score_val,chi_score_LM,pval = Wlinear_regression(df2)
chi_score_PM,pval = W_Polyfit(df2,3)
print 'Chi2: Linear model = {} Poly Model = {}'.format(chi_score_LM,chi_score_PM)

Chi2: Linear model = [ 328.1583208] Poly Model = 246.47135876


Q4] Use Chi^2 to find the polynomial order of the model that best fits the data? What one word describes why this value is not suitable for producing a predictive model.

In [41]:
#Validation

order_list = 10
best_Chi2 = 500

for i in range(order_list):
    chi_score,pval = W_Polyfit(df2,i)
    if (chi_score < best_Chi2):
        best_Chi2 = chi_score
        print "Degree {} has Chi2 of {}".format(i,int(best_Chi2))

Degree 1 has Chi2 of 426
Degree 2 has Chi2 of 245
Degree 8 has Chi2 of 241
Degree 9 has Chi2 of 229


A4]. 9. Overfitting.